runtime/src/main/java/org/corfudb/protocols/logprotocol/SMREntry.java

```diff
@@ -120,27 +124,33 @@ void deserializeBuffer(ByteBuf b, CorfuRuntime rt) {
         SMRMethod = new String(methodBytes);
         byte serializerId = b.readByte();
         byte numArguments = b.readByte();
-        Object[] arguments = new Object[numArguments];
+        List<Object> arguments = new ArrayList<>();
 
         if (!opaque) {
             serializerType = Serializers.getSerializer(serializerId);
         } else {
             this.serializerId = serializerId;
         }
 
+        List<Pair<ByteBuf, Integer>> objBufPairs= new ArrayList<>();
         for (byte arg = 0; arg < numArguments; arg++) {
             int len = b.readInt();
             ByteBuf objBuf = b.slice(b.readerIndex(), len);
+            objBufPairs.add(new ImmutablePair<>(objBuf, len));
+            b.skipBytes(len);
+        }
+
+        objBufPairs.parallelStream().forEachOrdered(objBufPair -> {
```

Comment:
The issue with this, is that it will give you parallelisim within a single transaction, but not across transactions. For write heavy patterns, this solution won't improve the sync performance. I suggest you try this at a higher level, for example:

```java

    @Override
    public Stream<SMREntry> streamUpTo(long maxGlobal) {
        return streamView.streamUpTo(maxGlobal)
                .filter(m -> m.getType() == DataType.DATA)
                .filter(m -> m.getPayload(runtime) instanceof ISMRConsumable
                        || m.hasCheckpointMetadata())
                .map(this::dataAndCheckpointMapper)
                .flatMap(List::stream);
    }

```

Reply:
@Maithem this solution works for customers who are batching their writes in a single transaction and this experiment was also conducted to see if this approach can help address the current read performance problems we are facing in that context. I would recommend us taking a 2 phase approach here i.e. push this CL to address batch writes customer which is a performance blocker now and then later as you suggested we can take this up to a higher level. We are also running really short on time as we have less than week to bake this change.